### PR TITLE
docs: use autosummary generated docs, add nitpicky mode

### LINKED ISSUE #1766: improve documentation search results
## 📚 Documentation
Currently searching docs on duckduckgo and google results in poor search results.
DuckDuckGo: search precision metrics show source code result first while third one is a correct page.

Screen Shot 2021-03-10 at 22 57 27

Google: nearly same to google as well tho 2nd one is correct page.

Screen Shot 2021-03-10 at 22 58 08

For Algolia Docsearch: source code are shown first then correct ones.

Screen Shot 2021-03-10 at 23 04 31

---

This might be due to title of the page generated by `_modules` folder.
To improve this, we have the following options
1. use `autosummary` and generate the docs for only classes or functions or both just like [nn.Module](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module)
2. put heading 2 to every classes and functions before `autoclass`, `autofunction`
3. make `_modules` not to be indexed by search engines and algolia

Here's my suggestions:
- For search engine results, I think 1 > 2 > 3. IMO generating long docstring classes and functions to separate page would result better search result by search engines and docs readability (sometimes I got lost in long docstrings and methods). **generating such pages have a downside, we can't specify `:members:` in `.. autoclass` manually** since it will be already included in template. See [pytorch template](https://github.com/pytorch/pytorch/blob/master/docs/source/_templates/classtemplate.rst).
- For algolia docsearch, 1 == 2 > 3. Algolia docsearch searches h1-h6, so adding heading to every classes and functions can result better in-docs search.

What do you think ? @vfdev-5 @sdesrozis

### PR DESCRIPTION
**Pull request by ydcjeff**
Fixes #1766 

https://deploy-preview-1780--pytorch-ignite-preview.netlify.app

Description:
- use autosummary generated docs for {engine,handlers,distributed,metrics,contrib/metrics,contrib/handlers}.rst
- fix math equation rendering by pinning katex at 0.13.0
- add nitpicky mode to check cross reference internal links (`-n` in docs.yml)
- fix the cross reference according to nitpicky mode
- ignore some external package links which don't exist
- add missing docstrings

Check list:

- [ ] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

File: docs/source/contrib/engines.rst [hunk 1]
```diff
@@ -4,29 +4,26 @@ ignite.contrib.engines
...
-
-Truncated Backpropagation Throught Time
----------------------------------------
-
-.. automodule:: ignite.contrib.engines.tbptt
-   :members:
-
-Helper methods to setup trainer/evaluator
------------------------------------------
+    :toctree: ../generated
```
Comment (ydcjeff) [inline docs/source/contrib/engines.rst:27]: @vfdev-5 should this page reverted to previous version?

Comment (vfdev-5) [reply docs/source/contrib/engines.rst:27]: Yes, I liked the previous version.

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -17,7 +17,7 @@ jobs:
         with:
           python-version: 3.7
 
-      - run: sudo npm install katex@latest -g
+      - run: sudo npm install katex@0.13.0 -g
```
Comment (vfdev-5) [inline .github/workflows/docs.yml:20]: Can't we use the latest, what is the reason ?

Comment (ydcjeff) [reply .github/workflows/docs.yml:20]: I am afraid it breaks the rendering.

Comment (vfdev-5) [reply .github/workflows/docs.yml:20]: Yes, this weird...

File: docs/source/contrib/metrics.rst [hunk 3]
```diff
@@ -34,35 +37,20 @@ Complete list of metrics:
...
-
-.. autoclass:: MedianAbsoluteError
-
-.. autoclass:: MedianAbsolutePercentageError
-
-.. autoclass:: MedianRelativeAbsoluteError
-
-.. autoclass:: R2Score
-
-.. autoclass:: WaveHedgesDistance
```
Comment (trsvchn) [inline docs/source/contrib/metrics.rst:68]: So, now, all these classes will have their one pages, and `docs/source/contrib/metrics.rst` will just hold a full list of them, right?

Comment (ydcjeff) [reply docs/source/contrib/metrics.rst:68]: Yes, they will be in generated/ dir when docs building

Comment (trsvchn) [reply docs/source/contrib/metrics.rst:68]: Thanks!

**Review by trsvchn — APPROVED** [review]

**Comment by trsvchn** [comment]
Thanks for the PR @ydcjeff LGTM! I left a comment about small nit)

File: docs/source/faq.rst [hunk 4]
```diff
@@ -36,7 +36,7 @@ As a consequence, the following code is correct too:
     evaluator.add_event_handler(Events.COMPLETED, best_model_saver, {'mymodel': model})
 
 
-More details :ref:`Events and Handlers`.
+More details :ref:`Events and Handlers:`.
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply docs/source/faq.rst:39]: Yes, it is a ref link now to https://github.com/pytorch/ignite/pull/1780/files#diff-e773146a9c09422b372307817044bc48d4277cbfdbb3bc0e3de6015824ab55e0R111

[the remark this answers is not in the document]
Comment (trsvchn) [reply docs/source/faq.rst:39]: I see, thanks!

**Review by vfdev-5 — APPROVED** [review]
Thanks a lot @ydcjeff for the PR and @trsvchn for the review !

**Comment by ydcjeff** [comment]
A corresponding PR to `docsearch-configs` : https://github.com/algolia/docsearch-configs/pull/3796

**Pull request merged**